Update example-sim to current standards

fgamador opened the pull request:
As an exercise to help me learn SceneryStack, I updated example-sim to match what seem to be the current standards. In the hopes that this will help future new devs, here are my changes. Please adjust as needed to fix things I got wrong. Here's the summary:

- Converted to TypeScript.
- Switched from using merge to optionize. (Not sure I handled ExampleSimConstants.SCREEN_OPTIONS correctly.)
- Added and used ExampleSimColors file.
- Added/updated other files to match simula-rasa. (I generated a new, empty example-sim using create-sim and diffed against it.)
- Fixed lint errors, including removing javadoc-style @param and @return keywords. Not sure what should replace these.

Note that while I tried to match existing formatting, I don't have a WebStorm license, so I didn't run an auto-formatter. Is there a way to do that without buying WebStorm?

fgamador committed:
Update files to look as if they were initially generated from simula-rasa

fgamador committed:
Add doc files as if from simula-rasa

fgamador committed:
Update MagnetsModel.ts to follow current TypeScript conventions (per simula-rasa)

fgamador committed:
Update BarMagnet.ts to follow current TypeScript conventions (per simula-rasa)

fgamador committed:
Update magnets view files to follow current TypeScript conventions (per simula-rasa)

fgamador committed:
Convert MagnetsControlPanel to use new-style options

fgamador committed:
Add param types and misc tidying

fgamador committed:
Fix formatting

fgamador committed:
Convert Particle to use new-style options

fgamador committed:
Add param types and misc tidying

fgamador committed:
Convert ParticleNode to add types and to use new-style options

fgamador committed:
Add param types and misc tidying

fgamador committed:
Add param types and misc tidying

fgamador committed:
Declare model fields

fgamador committed:
Declare model fields

fgamador committed:
Use EmptySelfOptions

fgamador committed:
Remove unused options

fgamador committed:
Misc tidying

fgamador committed:
Update MagnetsScreen.ts to use optionize

fgamador committed:
Update ParticlesScreen.ts to use optionize

fgamador committed:
Declare model fields

fgamador committed:
Declare model fields

fgamador committed:
Declare view fields

fgamador committed:
Use ExampleSimColors.screenBackgroundColorProperty

fgamador committed:
Use ExampleSimColors, plus misc tidying

fgamador committed:
Fix lint errors

fgamador committed:
Fix bad tandem create line

fgamador committed:
Better comment

fgamador committed:
Add dummy doc files

pixelzoom commented:
Adding this to the Developer Priorities board, to be discussed at an upcoming PhET developer meeting.

Thanks for the work here @fgamador, this looks very nice.

Before accepting this pull requests, we'll need to revisit is whether PhET wants example-sim to be in TypeScript.  The current answer was "no". Back when I doing a lot of proactive conversion from JavaScript to TypeScript (2022?), PhET decided _not_ to convert to example-sim to TypeScrpt because:

(1) example-sim is used for programming tasks during PhET's hiring process, where JavaScript experience is a prerequisite, but TypeScript is not.  Requiring TypeScript experience would significantly reduce the number of qualified candidates.

(2) For the open source community, TypeScript may be an additional barrier to getting involved with PhET code. The PhET tool chain certainly requires TypeScript, but it's still possible to write a sim entirely in JavaScript.

I think these points are both worth revisiting.

fgamador commented:
Okay, makes sense. I was wondering whether there was a reason example-sim is still JavaScript, but then I saw TypeScript requirements elsewhere. Whether or not you accept this, I found the conversion a valuable exercise, especially since I haven't used TypeScript before, and my JavaScript is pretty rusty.

marlitas commented:
Hi @fgamador, we decided that we are ready to convert example-sim to typescript and are very grateful for your work here! I will be reviewing the code in the next few days and will send any questions or revisions your way by early next week so we can merge your pull request in.

fgamador commented:
Okay, sound good. Glad to be able to help.

marlitas commented on js/particles/view/ParticlesScreenView.ts:
```diff
@@ -15,17 +15,23 @@ import TimeControlNode from '../../../../scenery-phet/js/TimeControlNode.js';
-class ParticlesScreenView extends ScreenView {
+export default class ParticlesScreenView extends ScreenView {
+
+  private model: ParticlesModel;
```
this should be readonly

marlitas commented on js/particles/view/ParticlesScreenView.ts:
```diff
@@ -15,17 +15,23 @@ import TimeControlNode from '../../../../scenery-phet/js/TimeControlNode.js';
+  private model: ParticlesModel;
+
+  // the parent for all ParticleNode instances
+  private particlesNode: Node;
```
This seems to me like it should also be readonly.

marlitas commented on js/particles/view/ParticlesScreenView.ts:
```diff
@@ -15,17 +15,23 @@ import TimeControlNode from '../../../../scenery-phet/js/TimeControlNode.js';
+   * model - the top-level model for this screen
    */
-  constructor( model ) {
+  public constructor( model: ParticlesModel ) {
```
You can actually declare and define the class field here like so: 
```
public constructor( private readonly model: ParticlesModel )
```

marlitas replied:
This would reduce code in the class field declarations as well as the body of the constructor

fgamador replied:
Wow, I didn't know you could do that. How do you decide whether to do so or not? Seems like the fields should all be one way of the other.

fgamador replied:
Okay, done.

marlitas reviewed (changes requested):
Thanks again for working on this conversion to typescript! I provided some feedback on the pull request. Please let me know if you have any questions. 

There are two pieces of feedback to keep an eye out for across all of the changes:

1. Ensure that git history is maintained. Files should be tracked as a rename instead of a delete and new file creation.

  **current files reporting as deleted:**
  - example-sim-main
  - MagnetsScreen
  - BarMagnet
  - ParticlesScreen
  - Particle
  - ParticleNode

2. Example sim should not worry about any PhET-iO support. In the changes you submitted, this would just include the removal of any tandems that have been added.

Once these requests are addressed we can move forward with next steps.

fgamador commented on js/example-sim-main.js:
I'm not sure how best to handle this. It looks like git does not actually track file renames, it only infers them based on how similar the file content is. If I do the rename in a separate commit as you suggest, then that makes it easy for git to do this inference. However, if you then squash my changes when merging in my PR, doesn't that lose the fact that the rename is a separate commit? Or will you keep all my individual commits?

Also, where can I find the `copy-history-to-different-repo` script?

marlitas replied:
I will be honest, we haven't done much work with pull requests here, so there might be some form of navigating as we go. However, based on the research I've done it looks like we can merge without squashing the commits. This is in fact the default merge option that [github provides. ](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges) Which I think we should start with.

So short answer, yes we should retain your individual commits during the merge.

fgamador replied:
Okay, I rebuilt the branch with a separate commit with all the renames. That required a rebase and a force push, which obsoleted all the existing commits. I didn't see a way around that.

fgamador commented on js/example-sim-main.ts:
```diff
@@ -0,0 +1,47 @@
+  const titleStringProperty = ExampleSimStrings[ 'example-sim' ].titleStringProperty;
+
+  const screens = [
+    new MagnetsScreen( { tandem: Tandem.ROOT.createTandem( 'magnetsScreen' ) } ),
```
I included the tandem stuff because create-sim adds it, and I wanted example-sim to match that. I guess we want example-sim to diverge from simula-rasa in this respect?

fgamador commented:
Thank you, @marlitas, for your thorough review. I have one other item I've been wondering about, which is the jsdoc annotations. I removed (most of?) them because when I ran lint from grunt, it complained about them. But now I'm unsure that that was the right thing to do. They are partly redundant to declarations in TypeScript, such as @public and parameter types, but maybe not entirely. Do you have a standard on how TypeScript should be commented?

fgamador committed:
Update per PR review comments

fgamador committed:
Fix call to optionize4

fgamador commented on js/magnets/MagnetsScreen.ts:
```diff
@@ -0,0 +1,55 @@
+
+  public constructor( providedOptions?: MagnetsScreenOptions ) {
+
+    const options = optionize<MagnetsScreenOptions, SelfOptions, ScreenOptions>()(
```
`optionize4` is complaining for me. ScreenOptions requires tandem. When I use it like this:

```
    const options = optionize4<MagnetsScreenOptions, SelfOptions, ScreenOptions>()(
      {}, ExampleSimConstants.SCREEN_OPTIONS,
      {
        name: ExampleSimStrings.screen.magnetsStringProperty,
        homeScreenIcon: createScreenIcon()
      },
      providedOptions );
```

It (meaning VS Code with the eslint plugin) says this about the default-values object-literal param:

"Argument of type '{ name: LocalizedStringProperty; homeScreenIcon: ScreenIcon; }' is not assignable to parameter of type 'OptionizeDefaults<EmptySelfOptions, ScreenOptions, EmptySelfOptions>'.
  Property 'tandem' is missing in type '{ name: LocalizedStringProperty; homeScreenIcon: ScreenIcon; }' but required in type 'Required<Omit<Pick<ScreenOptions, "tandem">, never>>'.ts(2345)
PhetioObject.ts(134, 3): 'tandem' is declared here."

marlitas replied:
Ahh great, there are a couple of ways you should be able to address this. You can adjust your code block with either of the following options. I'm going to tag @samreid and @pixelzoom here in case they have any further insight onto how we would like open source contributors to handle tandems.

Option 1:
```
const options = optionize4<MagnetsScreenOptions, SelfOptions, StrictOmit<ScreenOptions, 'tandem'>>()(
      {}, ExampleSimConstants.SCREEN_OPTIONS,
      {
        name: ExampleSimStrings.screen.magnetsStringProperty,
        homeScreenIcon: createScreenIcon()
      },
      providedOptions );
```

Option 2:
```
const options = optionize4<MagnetsScreenOptions, SelfOptions, ScreenOptions>()(
      {}, ExampleSimConstants.SCREEN_OPTIONS,
      {
        name: ExampleSimStrings.screen.magnetsStringProperty,
        homeScreenIcon: createScreenIcon(),
        tandem: Tandem.OPT_OUT
      },
      providedOptions );
```

fgamador replied:
I picked "tandem: Tandem.OPT_OUT" as it seems marginally more clear on what's going on.

fgamador commented on js/magnets/model/MagnetsModel.ts:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2013-2021, University of Colorado Boulder
```
Probably just luck, per my comment above. I thought I could fix this by using `git mv`, but according to the interwebs, `git mv` is just a shorthand for delete and add.

marlitas commented:
>  I removed (most of?) them because when I ran lint from grunt, it complained about them. But now I'm unsure that that was the right thing to do.

@fgamador, yes that is correct! your jsdoc annotations looked correct. We do not document the redundant typescript information in jsdoc, since it can quickly become a nuisance to maintain and update two spots at once. Generally the addition information jsdoc will provide is specific to the implementation or context of a function, parameter, class field, etc. 

Overall your jsdoc looked as I would expect and no needed changes were apparent there.

marlitas commented:
@fgamador so sorry! I didn't even notice those. I'll take a look later today and respond. thanks for the ping!

marlitas commented:
@fgamador I believe I responded to all of your questions, but please let me know if I missed any!

fgamador committed:
Rename .js files as .ts files

fgamador committed:
Update and add .ts files to match simula-rasa

fgamador commented:
Okay, I think I've addressed all your comments, @marlitas. Sorry for the slow response; I've been out of town. And I'm heading out again, so it'll be another week before I can respond again.

fgamador committed:
Update per PR review comments

marlitas commented:
@fgamador no worries. It will take a moment for me to look through it and review. Thank you so much!

marlitas commented:
@fgamador we were successfully able to merge everything in and it's all working and looking as expected. Thanks again for all your work here!
